[TEAM 02] 연 & 시온 - OAuth 구현과 이슈, 코멘트, 레이블, 마일스톤 관련 기능 구현 완료했습니다.

kimnayeon0108 opened the pull request:
안녕하세요! 백엔드 연과 시온입니다. 

**궁금한 점**
spring.jpa.hibernate.ddl-auto을 create로 설정 하였습니다. 
테스트를 하기위해 data.sql에 더미데이터를 추가했습니다. 그런데 DB에 있는 데이터를 초기화 하기 위해 spring.jpa.defer-datasource-initialization=true 설정을 하지 않으면 기존 데이터때문에 에러가 발생합니다. 
ddl 설정을 create로 하면 자동으로 테이블이 Drop 되고 재생성되는 걸로 알고 있는데 initialization 설정이 왜 필요한지 의문입니다. 찾아보니 Session factory 라는 것이 있는데 혹시 이와 관련된건지 궁금합니다...!



## 구현한 기능
**OAuth**
- GitHub 로그인
- JWT
- FE, iOS와 GitHub 로그인 연동 

**이슈**
- 이슈 전체 조회
- 이슈 상세 조회 
- 이슈 등록
- 선택한 이슈 닫기, 열기
- 이슈 제목, 담당자, 레이블, 마일스톤 변경

**코멘트**
- 코멘트 등록
- 코멘트 이모지 추가
- 코멘트 변경

**레이블**
- 전체 레이블 조회
- 레이블 변경
- 레이블 추가
- 레이블 삭제 

**마일스톤**
- 전체 마일스톤 조회
- 마일스톤 개수 조회
- 마일스톤 추가
- 마일스톤 변경
- 마일스톤 삭제 

리뷰해주셔서 감사합니다.

kimnayeon0108 committed:
refactor : 정적 팩토리 메소드 대신 매개변수로 도메인 객체를 받는 생성자로 대체

kimnayeon0108 committed:
refactor : 메소드 변경에 따른 코드 수정

Commit:
Merge pull request #20 from Sonjh1306/BE/feature/issue

[#15] 이슈 전체보기 구현

Commit:
[#14] feat : 깃헙 로그인시 회원가입

kimnayeon0108 committed:
chore : 기능 테스트를 위한 더미데이터 변경

2번 issue의 is_open 속성 false로 변경

kimnayeon0108 committed:
feat : findById 메소드 추가

kimnayeon0108 committed:
feat : Issue 관련 Response DTO 클래스들의 필드 중복의 이유로 AbstractIssueResponse 생성

이슈 상세 보기를 위한 DetailIssueResponse 생성

kimnayeon0108 committed:
feat : 마일스톤이 갖고 있는 이슈들의 개수 반환 기능 추가, getTotalIssueCount, getOpenIssueCount

kimnayeon0108 committed:
test : 마일스톤이 갖고 있는 이슈들의 개수 반환 기능 테스트 추가

kimnayeon0108 committed:
feat : MilestoneDetailResponse 추가

이슈 상세보기를 위한 마일스톤 Response DTO

kimnayeon0108 committed:
feat : Response DTO wrapping 용도의 DataResponse 분리

Response List를 래핑하는 ListDataResponse
DTO 객체 하나만 래핑하는 DataResponse

kimnayeon0108 committed:
feat : IssueNotFoundException 예외 정의

kimnayeon0108 committed:
feat : getDetailIssueResponse 메소드 추가

kimnayeon0108 committed:
feat : 이슈 상세보기 기능 추가, showDetailIssue 메소드

kimnayeon0108 committed:
feat : wrapping dto용 디렉토리 분리

kimnayeon0108 committed:
feat : dto 디렉토리 하위에 issue 디렉토리 생성

issue 관련 dto용 디렉토리

kimnayeon0108 committed:
feat : domain 클래스에 Getter 추가 (Comment, CommentEmoji, Emoji)

kimnayeon0108 committed:
feat : EmojiResponse 추가

kimnayeon0108 committed:
feat : CommentResponse 추가

CommentEmoji 도메인 객체 리스트를 EmojiResponse dto 리스트로 변환하는 기능 추가

kimnayeon0108 committed:
chore : import문 수정

kimnayeon0108 committed:
chore : import문 수정

kimnayeon0108 committed:
chore : comment, emoji, comment_emoji 더미데이터 추가

Commit:
[#14] feat : 이미 회원가입 되어 있으면 회원가입 하지 않는다.

kimnayeon0108 committed:
refactor : final 키워드 추가

Commit:
[#14] refactor : ApiResult 추가

Commit:
[#14] refactor : oauth 프로필 설정 추가

kimnayeon0108 committed:
feat : 이슈에 해당하는 마일스톤이 없는 경우 null을 Optional로 감싸서 값을 넣도록 구현

kimnayeon0108 committed:
chore : 테스트를 위한 더미 데이터 수정

Commit:
Merge pull request #23 from Sonjh1306/BE/feature/issue-detail

[#15] 이슈 상세보기 구현

kimnayeon0108 committed:
chore : import문 수정

kimnayeon0108 committed:
feat : repository 상속 인터페이스 변경

CrudRepository에서 JpaRepository로 변경

kimnayeon0108 committed:
feat : Request DTO클래스 역할의 IssueRequest 생성

kimnayeon0108 committed:
feat : IssueRequest DTO 클래스 생성

kimnayeon0108 committed:
feat : 기본 생성자와 content, file을 매개변수로 받는 생성자 추가

kimnayeon0108 committed:
chore : import문 수정

kimnayeon0108 committed:
chore : 에러 메시지 수정

kimnayeon0108 committed:
feat : NotFoundException 생성

kimnayeon0108 committed:
feat : LabelNotFoundException 생성

kimnayeon0108 committed:
feat : MilestoneNotFoundException 생성

kimnayeon0108 committed:
feat : UserNotFoundException 생성

kimnayeon0108 committed:
feat : LabelRepository 생성, findById 메소드 추가

kimnayeon0108 committed:
feat : MilestoneRepository 생성, findById 메소드 추가

kimnayeon0108 committed:
feat : UserRepository 생성, findById 메소드 추가

kimnayeon0108 committed:
feat : 추가적인 repository 의존성 주입, addIssue 메소드 추가

IssueRequest를 Issue로 변환할 때 필요한 UserRepository, LabelRepository, MilestoneRepository 의존성 추가
Label의 id값 리스트로 Label 리스트 생성하는 getLabels 메소드 추가
assignee의 id값 리스트로 User 리스트 생성하는 getAssignees 메소드 추가

Commit:
[#14] feat : web, ios 깃헙로그인 요청을 나눔

Commit:
[#14] refactor : aop 적용해서 공통 로깅

Commit:
[#14] refactor : 깃헙 로그인 관련 프로퍼티를 따로 관리한다.

Commit:
[#14] refactor : 유저정보에서 password 제거

Commit:
[#14] refactor : aop 메서드명 구체화

Commit:
[#14] chore : oauth.properties gitignore 추가

Commit:
merge : dev/BE into BE/feature/oauth

Commit:
[#14] fix : merge 후 누락 부분 해결 및 소셜로그인 테스트 위해 data 변경

kimnayeon0108 committed:
feat : Id 생성 전략 설정, content, file, writer를 매개변수로 갖는 생성자와 addIssue 메소드 추가

kimnayeon0108 committed:
feat : Id 생성 전략 설정, title, writer, isOpen을 매개변수로 갖는 생성자와 addMilestone 메소드 추가

kimnayeon0108 committed:
feat : toIssue 메소드와 toComment 메소드 추가

kimnayeon0108 committed:
feat : 기본 생성자와 모든 필드를 매개변수로 갖는 생성자 추가

kimnayeon0108 committed:
feat : CommentRepository 추가

kimnayeon0108 committed:
feat : IssueAssigneeRepository 추가

kimnayeon0108 committed:
feat : IssueLabelRepository 추가

kimnayeon0108 committed:
feat : 새로운 Issue 생성 기능 구현

IssueRequest DTO를 통해 Comment, Issue, IssueLabel, IssueAssignee를 생성하고 DB에 저장

kimnayeon0108 committed:
feat : Response DTO wrapping 용도의 ApiResult 클래스 생성

기존의 DataResponse와 ListDataResponse 삭제

kimnayeon0108 committed:
feat : return 타입 변경

kimnayeon0108 committed:
feat : return 타입 변경

kimnayeon0108 committed:
refactor : 제거된 필드와 관련된 코드 삭제

kimnayeon0108 committed:
refactor : IssueService 분리

CommentService, LabelService, MilestoneService, UserService를 생성해서 IssueService가 수행하던 역할을 각각에 맞는 역할 수행하도록 서비스단 분리함

kimnayeon0108 committed:
Merge branch 'dev/BE' into BE/feature/issue-create

Commit:
Merge pull request #27 from Sonjh1306/BE/feature/issue-create

[#15] 이슈 추가하기 구현 완료

Commit:
[#14] feat : jwt에 user id만 담고 로그인 처리

Commit:
[#14] chore : 패키지 구조 변경

Commit:
[#14] refactor : globa exception handler 추가

Commit:
Merge branch 'dev/BE' into BE/feature/oauth

kimnayeon0108 committed:
Merge pull request #26 from Sonjh1306/BE/feature/oauth

깃헙 로그인으로 jwt 발급하고 요청에 로그인 처리

kimnayeon0108 committed:
refactor : IssueRequest DTO에서 userId 제외

Commit:
chore : dev/BE 에 기능 합친 후 conflict 해결

kimnayeon0108 committed:
feat : 인터셉터를 통해 JWT 토큰에서 얻은 Id값을 매개변수로 받도록 변경함

메소드단에 LoginRequired 어노테이션 추가
UserId 어노테이션을 단 userId 매개변수 추가

kimnayeon0108 committed:
refactor : addIssue 메소드에 userId 매개변수 받도록 변경함

이에 따른 코드 수정

kimnayeon0108 committed:
refactor : import문 수정

kimnayeon0108 committed:
Merge pull request #28 from Sonjh1306/BE/fix/conflict

chore : dev/BE 에 기능 합친 후 conflict 해결

Commit:
Merge pull request #29 from Sonjh1306/BE/feat/issue-oauth-matching

chore: JWT에서 userId를 받아서 이슈 생성시 사용하도록 변경

kimnayeon0108 committed:
feat : closeIssues 메소드 추가

kimnayeon0108 committed:
feat : IssueIdsRequest DTO 클래스 추가

kimnayeon0108 committed:
feat : Issue 엔티티 클래스에 close 메소드 추가

kimnayeon0108 committed:
feat : closeIssues 메소드 추가

Issue Id를 Request DTO를 통해 얻어서 해당되는 issue들을 close 하는 기능 추가

kimnayeon0108 committed:
feat : IllegalIssueStatusException 예외 추가

kimnayeon0108 committed:
chore : 더미 데이터 수정

kimnayeon0108 committed:
feat : open 메소드 추가, 열린 이슈나 닫힌 이슈가 아니면 예외 발생하도록 함

kimnayeon0108 committed:
feat : openIssues 메소드 추가

예외 발생시키는 로직 엔티티 클래스로 옮김

kimnayeon0108 committed:
feat : openIssues 메소드 추가

kimnayeon0108 committed:
feat : NotFoundException 과 IllegalIssueStatusException 핸들링 추가

kimnayeon0108 committed:
feat : IssueTitleRequest DTO 클래스 추가

kimnayeon0108 committed:
feat : fixTitle 메소드 추가

kimnayeon0108 committed:
feat : modifyTitle 메소드 추가

kimnayeon0108 committed:
feat : modifyTitle 메소드 추가

이슈의 제목 변경하는 기능

kimnayeon0108 committed:
feat : IssueAssigneeIdsRequest DTO 클래스 추가

이슈 담당자 수정시 Request DTO

kimnayeon0108 committed:
feat : findByIssueId 메소드 추가

kimnayeon0108 committed:
feat : fixIssueAssignees 메소드 추가

kimnayeon0108 committed:
feat : modifyIssueAssignees와 deleteIssueAssignees 메소드 추가

kimnayeon0108 committed:
feat : modifyAssignees 메소드 추가

kimnayeon0108 committed:
test : findByIssueId 메소드 기능 테스트 추가

kimnayeon0108 committed:
feat : modifyAssignees 메소드 추가

kimnayeon0108 committed:
feat : IssueLabelIdsRequest DTO 클래스 추가

kimnayeon0108 committed:
feat : findByIssueId 메소드 추가

kimnayeon0108 committed:
feat : fixIssueLabels 메소드 추가

kimnayeon0108 committed:
refactor : 내용 변경하는 메소드명 replace로 변경

kimnayeon0108 committed:
feat : modifyLabels 메소드 추가

kimnayeon0108 committed:
feat : modifyIssueLabels, deleteIssueLabels 메소드 추가

kimnayeon0108 committed:
feat : modifyLabels 메소드 추가

이슈의 레이블 변경 기능

kimnayeon0108 committed:
chore : import문 수정

kimnayeon0108 committed:
feat : replaceMilestone 메소드 추가

kimnayeon0108 committed:
feat : IssueMilestoneRequest DTO 클래스 추가

kimnayeon0108 committed:
feat : modifyMilestone 메소드 추가

kimnayeon0108 committed:
feat : modifyMilestone 메소드 추가

이슈의 마일스톤 변경하는 기능

kimnayeon0108 committed:
refactor : import문 수정

kimnayeon0108 committed:
refactor : import문 수정

kimnayeon0108 committed:
refactor : findOne 매개변수 id로 변경

kimnayeon0108 committed:
feature : CommentRequest DTO 클래스 추가

kimnayeon0108 committed:
feature : addComment 메소드 추가

kimnayeon0108 committed:
feature : save 메소드 추가

kimnayeon0108 committed:
feature : addComment 메소드 추가

kimnayeon0108 committed:
feature : createComment 메소드 추가

코멘트 등록 기능

kimnayeon0108 committed:
feature : 엔티티 클래스에 isDeleted 필드 추가

soft delete 구현을 위한 필드 추가

kimnayeon0108 committed:
feature : getAllIssueResponses 메소드 수정, 이슈 생성 혹은 마일스톤 변경시 milestoneId가 0인 경우 처리, validateDeleted 메소드 추가

이슈 전체 조회시 isDeleted가 false인 것만 반환하도록 수정
milestoneId가 0인 경우 매핑된 milestone null로 처리
이슈 관련 사항 변경시 validateDeleted 로 삭제된 이슈인지 확인하는 코드 추가

kimnayeon0108 committed:
refactor : getMilestone, isMilestoneEmpty 메소드 이동

IssueService에서 MilestoneService로 로직 이동

kimnayeon0108 committed:
feature : IllegalStatusException 사용자 예외 이름 변경

IllegalIssueStatusException 에서 IllegalStatusException 으로 변경

kimnayeon0108 committed:
refactor : 예외 이름 변경에 따른 코드 수정

kimnayeon0108 committed:
feat : CommentEmojiRepository 생성

kimnayeon0108 committed:
test : CommentEmojiRepository의 findByCommentId 메소드가 잘 동작하는지 확인하는 테스트 추가

kimnayeon0108 committed:
feat : 엔티티 클래스에 delete 기능 추가

kimnayeon0108 committed:
feat : deleteIssue 메소드 추가

soft delete 방식으로 이슈 삭제
이슈에 포함된 코멘트도 soft delete 되도록 구현
코멘트와 연결된 CommentEmoji 삭제
해당 Issue와 연결된 IssueAssignees와 IssueLabels 삭제

kimnayeon0108 committed:
feat : getTotalIssueCount, getOpenIssueCount 메소드 수정

delete된 issue는 포함하지 않도록 변경

kimnayeon0108 committed:
feat : saveAll, deleteCommentEmojis 메소드 추가

kimnayeon0108 committed:
refactor : 메소드명 변경에 따른 코드 수정

kimnayeon0108 committed:
refactor : 접근제어자 변경

kimnayeon0108 committed:
feat : 마일스톤 상태 검사 로직 추가

kimnayeon0108 committed:
refactor : 접근제어자 수정

kimnayeon0108 committed:
feat : deleteIssue 메소드 추가

이슈 삭제 기능

kimnayeon0108 committed:
feat : Issue 클래스의 getComment 메소드 수정

삭제되지 않은 comment만 반환되도록 수정

kimnayeon0108 committed:
feat : CommentEmojiRequest DTO 클래스 추가

kimnayeon0108 committed:
feat : EmojiRepository 클래스 추가

kimnayeon0108 committed:
feat : EmojiNotFoundException 사용자 예외 추가

kimnayeon0108 committed:
feat : CommentNotFoundException 사용자 예외 추가

kimnayeon0108 committed:
feat : 생성자 추가

kimnayeon0108 committed:
feat : addEmoji 메소드 추가

kimnayeon0108 committed:
feat : CommentController 클래스 생성, addEmojiOnComment 메소드 추가

kimnayeon0108 committed:
test : comment 가져오는 코드에 orElseThrow로 수정

kimnayeon0108 committed:
feat : replaceContent, replaceFile 메소드 추가

kimnayeon0108 committed:
feat : modifyComment 메소드 추가

kimnayeon0108 committed:
feat : modifyComment 메소드 추가

코멘트 수정 기능

kimnayeon0108 committed:
chore : 더미데이터 수정

kimnayeon0108 committed:
test : CommentRepositoryTest 테스트 클래스 생성

IssueRepositoryTest에 있던 코멘트 관련 테스트 이동

kimnayeon0108 committed:
feat : 이슈와 같이 조회되는 레이블 Response DTO 클래스명 변경

기존 LabelResponse에서 IssueLabelResponse로 변경

kimnayeon0108 committed:
refactor : 클래스명 변경에 따른 코드 수정

kimnayeon0108 committed:
feat : Label 전체조회를 위한 Response DTO 용도의 LabelResponse 클래스 생성

kimnayeon0108 committed:
feat : LabelRepository에 findAll 메소드 추가

isDeleted=false 인 것만 반환하도록 구현

kimnayeon0108 committed:
feat : LabelController 클래스 생성, showAllLabels 메소드 구현

kimnayeon0108 committed:
feat : LabelService에 getAllLabelResponses 메소드 추가

kimnayeon0108 committed:
test : LabelRepositoryTest 클래스 생성과 findAll 메소드 테스트 추가

kimnayeon0108 committed:
chore : 더미 데이터 추가

kimnayeon0108 committed:
feat : Issue 개별 조회시 삭제되지 않은 코멘트만 반환하도록 수정

kimnayeon0108 committed:
feat : Repository에서 데이터 조회시 isDeleted가 false인 것만 반환하도록 수정

kimnayeon0108 committed:
refactor : 불필요한 코드 제거

kimnayeon0108 committed:
test : IssueRepository의 findAll 메소드 변경에 따른 테스트 수정

kimnayeon0108 committed:
chore : 더미 데이터 수정

kimnayeon0108 committed:
Merge branch 'BE/feature/issue-modify' into BE/feature/label

kimnayeon0108 committed:
feat : LabelCountResponse DTO 클래스 추가

kimnayeon0108 committed:
chore : import문 수정

kimnayeon0108 committed:
feat : 레이블의 개수를 반환하는 getLabelCount 메소드 추가

kimnayeon0108 committed:
feat : LabelController에서 레이블의 개수를 보여주는 showLabelCount 메소드 추가

kimnayeon0108 committed:
refactor : 불필요한 코드 제거

kimnayeon0108 committed:
feat : LabelRequest에 따라 레이블의 필드 값을 변경하는 modifyLabel 메소드 추가

kimnayeon0108 committed:
feat : LabelRequest DTO 클래스 추가

kimnayeon0108 committed:
feat : LabelRepository의 findById 메소드 수정

삭제되지 않은 레이블만 반환하도록 변경

kimnayeon0108 committed:
feat : LabelService에 modifyLabel 기능 추가

kimnayeon0108 committed:
feat : LabelController에 ModifyLabel 기능 추가

LabelRequest에 따라 레이블을 수정하는 기능

kimnayeon0108 committed:
feat : 새로운 레이블 추가하기 구현

kimnayeon0108 committed:
feat : 레이블 삭제 구현

해당 레이블과 매핑된 IssueLabel도 함께 삭제

kimnayeon0108 committed:
refactor : 이슈 상세보기시 사용되는 Milestone DTO 클래스명 DetailIssueMilestoneResponse로 변경, Issue와 관련된 DTO 클래스들 issue 디렉토리로 이동

Commit:
Merge pull request #30 from Sonjh1306/BE/feature/issue-modify

[#15] 이슈 관련 수정 기능과 코멘트 API 구현

Commit:
Merge pull request #31 from Sonjh1306/BE/feature/label

[#17] 레이블 관련 API 구현

kimnayeon0108 committed:
feat : 마일스톤 전체 조회 기능 추가

kimnayeon0108 committed:
refactor : issueRepository 메소드 변경에 따라 불필요한 검증 로직 삭제

kimnayeon0108 committed:
feat : 마일스톤 개수 조회 기능 추가

kimnayeon0108 committed:
test : MilestoneRepository의 findOpenMilestones 메소드와 findClosedMilestones 메소드가 잘 동작하는지 확인하는 테스트 코드 추가

kimnayeon0108 committed:
feat : 마일스톤 추가 기능 구현

kimnayeon0108 committed:
feat : 마일스톤 변경 기능 구현

kimnayeon0108 committed:
refactor : 엔티티 클래스의 수정 기능 메소드명 edit으로 변경

kimnayeon0108 committed:
feat : 마일스톤 삭제 구현

kimnayeon0108 committed:
chore : 더미 데이터 수정

Commit:
[#14] refactor : 유저 검색시 조건을 이메일에서 login으로 변경

Commit:
[#14] chore : 프론트 cors 설정

Commit:
Merge pull request #32 from Sonjh1306/BE/feature/Milestone

[#18] 마일스톤 관련 기능 구현

Commit:
Merge branch 'dev/BE' into BE/chore/cors

kimnayeon0108 committed:
Merge pull request #33 from Sonjh1306/BE/chore/cors

Be/chore/cors

kimnayeon0108 committed:
feat : 이슈 개수 조회 기능 구현

kimnayeon0108 committed:
test : IssueRepository의 findOpenIssues 메소드와 findClosedIssues 메소드가 잘 동작하는지 확인하는 테스트 추가

kimnayeon0108 committed:
feat : 이슈, 레이블, 마일스톤 전체 개수 조회 기능 구현

kimnayeon0108 committed:
chore : 더미 데이터 수정

Commit:
Merge pull request #36 from Sonjh1306/BE/feature/issue-count

[#15] 이슈 개수 조회, 이슈, 레이블, 마일스톤 전체 개수 조회 구현 완료

ksundong commented:
리뷰가 늦어질 것 같아요!
조금만 기다려주세요~ 😃

ksundong commented on BE/issue_tracker/build.gradle:
```diff
@@ -21,10 +21,12 @@ repositories {
 	implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
 	implementation 'org.springframework.boot:spring-boot-starter-web'
-	compileOnly 'org.projectlombok:lombok'
+    implementation 'org.projectlombok:lombok:1.18.16'
```
스프링부트에서 버전관리 해줄거에요 한 번 알아보세요!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/config/AopConfig.java:
```diff
@@ -0,0 +1,44 @@
+
+@Slf4j
+@EnableAspectJAutoProxy
+@Aspect
```
오호 AOP를 사용하는군요!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/config/AopConfig.java:
```diff
@@ -0,0 +1,44 @@
+@Component
+public class AopConfig {
+
+    // 깃헙으로부터 받아온 인증코드를 출력한다.
```
자바독으로 작성해주는 편이 좋을 것 같아요~

ehdrhelr replied:
자바독을 적극 활용해 보겠습니다~!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/config/AopConfig.java:
```diff
@@ -0,0 +1,44 @@
+    // target 메소드의 동작 시간을 출력한다.
+    @Around("execution(* team02.issue_tracker.oauth.controller.OAuthController.*issue*(..))")
+    public Object jwtIssueTimeLog(ProceedingJoinPoint proceedingJoinPoint) throws Throwable{
+        long startTime = System.currentTimeMillis();
+
+        Object result = proceedingJoinPoint.proceed();
+
+        long endTime = System.currentTimeMillis();
+
+        log.info(proceedingJoinPoint.getSignature().getName() + " running time : {}ms", (endTime - startTime));
+
```
스프링에 StopWatch라는 것도 있는데 한 번 사용해보세요~

ehdrhelr replied:
적용해보겠습니다~!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/config/AopConfig.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    // 발행한 jwt를 출력한다.
+    @AfterReturning(value = "execution(* team02.issue_tracker.oauth.controller.OAuthController.*issue*(..))", returning = "authJwt")
```
전부 다 똑같은 곳을 가리키는 것 같은데, 특정 메서드를 지정해주는 편이 더 좋아보여요~

ehdrhelr replied:
오 적용해보겠습니다~

ehdrhelr replied:
넵 말씀듣고 메서드로 지정하고 커스텀 어노테이션도 사용해보았습니다!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/config/WebMvcConfig.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Override
+    public void addCorsMappings(CorsRegistry registry) {
+        registry.addMapping("/api/**")
+                .allowedOrigins("http://localhost:3000");
```
CORS에 줄 수 있는 옵션이 여러개 있는데요.
한 번 공부해보세요~

ehdrhelr replied:
더 공부해보겠습니다~!

kimnayeon0108 committed:
refactor : 필드명 변경

isDeleted 에서 deleted로 변경
isOpen에서 open으로 변경

kimnayeon0108 committed:
chore : 코드 줄맞춤, 불필요한 import문 제거

kimnayeon0108 committed:
refactor : Repository 메소드 수정

Query어노테이션으로 직접 쿼리문 작성한 부분 삭제하고 메소드명에 조건 명시하도록 변경

kimnayeon0108 committed:
chore : 더미 데이터 수정

kimnayeon0108 committed:
test : Repository의 메소드명 변경에 따른 테스트 코드 수정

Commit:
Merge pull request #41 from Sonjh1306/BE/refactor/Repository

[#40] refactor: boolean 필드명 변경, Repository 메소드 수정

kimnayeon0108 committed:
feat : 유저 정보 조회하는 기능 추가

Commit:
Merge pull request #43 from Sonjh1306/BE/feature/User

[#42] 유저 정보 조회하는 기능 추가

kimnayeon0108 committed:
feat : 이모지 전체 조회 기능 추가

kimnayeon0108 committed:
chore : 이모지 데이터 추가

Commit:
[#4] chore : front end local origin 허용 (CORS)

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/controller/CommentController.java:
```diff
@@ -0,0 +1,38 @@
+    @PostMapping("/{commentId}/emojis")
+    public ApiResult<String> addEmojiOnComment(@PathVariable Long commentId, @RequestBody CommentEmojiRequest commentEmojiRequest) {
+        commentService.addEmoji(commentId, commentEmojiRequest);
+        return ApiResult.ok();
```
`ApiResult.ok()`는 어떤 의미가 있는건가요?
단순 성공응답이라면 201이라던가 200응답을 주는 편이 좀 더 RESTful하지 않을까 싶어요.

ksundong replied:
이거 전부터 계속 피드백 드렸던 것 같은데~
RESTful한 API 설계를 어떻게 하는가는 한 번쯤 고민을 해보셨으면 좋겠어요!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/controller/CommentController.java:
```diff
@@ -0,0 +1,38 @@
+        return ApiResult.ok();
+    }
+
+    @GetMapping("/emojis")
+    public ApiResult<List<EmojiResponse>> showAllEmojis() {
+        return ApiResult.success(commentService.getAllEmojiResponses());
+    }
```
무엇의 이모지들을 다 가져오는건가요?
보여줄 수 있는 모든 이모지들을 서버에서 관리하는걸까요?

kimnayeon0108 replied:
@ksundong 네엡 이모지를 추가할때 모든 이모지를 보여줘야할 거 같아서 구현한 기능입니다~

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/controller/CommonController.java:
```diff
@@ -0,0 +1,36 @@
+import team02.issue_tracker.repository.MilestoneRepository;
+
+@RestController
+@RequestMapping("/api/count")
+public class CommonController {
```
count인데 common이에요. 뭔가 좀 어울리지 않는것 같은데요?

kimnayeon0108 replied:
@ksundong 
RequestMapping에 url을 `/api/common`으로 변경하고 showAllCount 메소드에 `/count` 로 변경했습니다!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/controller/CommonController.java:
```diff
@@ -0,0 +1,36 @@
+@RequestMapping("/api/count")
+public class CommonController {
+
+    private final IssueRepository issueRepository;
+    private final LabelRepository labelRepository;
+    private final MilestoneRepository milestoneRepository;
```
Repository를 Controller가 직접적으로 이용하는건 안좋은 패턴을 유도할 가능성이 높아보입니다.

kimnayeon0108 replied:
@ksundong 
CommonController와 UserController에서 service를 이용하도록 모두 변경했습니다~!

kimnayeon0108 replied:
@ksundong 이미 존재하는 서비스계층을 이용하면 됐는데 왜 이렇게 했는지.... 반성합니다ㅠ_ㅠ

ksundong replied:
그럴 수 있죠~ 다음엔 좀 더 신경써보세요~

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/controller/CommonController.java:
```diff
@@ -0,0 +1,36 @@
+        this.milestoneRepository = milestoneRepository;
+    }
+
+    @GetMapping
+    public ApiResult<CountResponse> showAllCount() {
+        int openIssueCount = issueRepository.findByOpenTrueAndDeletedFalse().size();
+        int closedIssueCount = issueRepository.findByOpenFalseAndDeletedFalse().size();
+        int labelCount = labelRepository.findByDeletedFalse().size();
+        int openMilestoneCount = milestoneRepository.findByOpenTrueAndDeletedFalse().size();
+        int closedMilestoneCount = milestoneRepository.findByOpenFalseAndDeletedFalse().size();
+
+        return ApiResult.success(new CountResponse(openIssueCount, closedIssueCount, labelCount, openMilestoneCount, closedMilestoneCount));
```
이렇게 해줘야만 하는 이유가 있었나요?
count만 필요하다면, count쿼리를 이용해봅시다.

그리고 여러번 쿼리하기 보다는 한 번에 쿼리해서 결과를 가져오는 편이 네트워크 비용을 아낄 수 있을거에요.

ksundong replied:
이런 경우엔 Repository의 기능을 이용하기 보다는
직접 Query를 작성하는 방식으로 구현하는 편이 좋습니다.

kimnayeon0108 replied:
@ksundong Repository의 count기능을 이용하도록 변경했습니다~ 
말씀해주신대로 hibernate의 soft delete 방식을 이용하도록 바꾸고 `@Where` 어노테이션으로 is_deleted=false 인 것은 조회하지 않도록 설정했습니다.
직접 쿼리를 작성하라는 이유가 메소드명이 너무 길고 복잡해져서 라고 생각했는데 soft delete 설정으로 count 메소드명이 크게 복잡해지지 않는거 같아서 Repository의 기능을 이용하는 쪽으로 구현했습니다....!

이부분도 더 생각하면서 구현 했어야 했는데.... 반성중입니다ㅠㅠ

ksundong replied:
ㅎㅎ 다음에 더 잘해보세요.

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/controller/IssueController.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    @DeleteMapping("/{issueId}")
+    public ApiResult<String> deleteIssue(@PathVariable Long issueId) {
```
이거 보면서 든 생각인데 String을 리턴시키는 것도 약간 의아한 형태입니다.

ksundong replied:
딱히 응답값이 없다면 Void를 사용하는 것도 좋아보여요.

kimnayeon0108 replied:
@ksundong ApiResult.ok 메소드를 호출했을 때 데이터의 String으로 "OK"를 반환해줘서 String으로 지정해주었습니다. Void는 예외 반환시 데이터에 null이 들어가기 때문에 예외 처리시에 Void를 사용하면 좋을거 같습니다...!

ksundong replied:
자 String으로 OK를 주는게 과연 좋은 형태일까요?
HTTP Response에 200응답코드는 뭐라고 부를까요?

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/controller/MilestoneController.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @DeleteMapping("/{milestoneId}")
+    public ApiResult<String> deleteMilestone(@PathVariable Long milestoneId) {
```
한가지 좋은 점은 API가 일관적인 형태여서 좋다고 생각이 됩니다.
사실 코드가 가장 잘 전달되려면 일관적인 형태를 갖추는게 제일 좋거든요!

나무랄 데 없이 일관적으로 잘 만들어주신 것 같아요!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/controller/UserController.java:
```diff
@@ -0,0 +1,30 @@
+@RequestMapping("/api/user")
+public class UserController {
+
+    private final UserRepository userRepository;
```
되도록이면 서비스 계층을 두는 것이 좋을 것 같습니다.
Presentation 계층이, Persistence 계층과 직접적인 연관관계를 맺는 것은 추후 확장 가능성을 고려해봤을 때, 불리한 선택이라고 생각이 되네요.

kimnayeon0108 replied:
@ksundong 네엡~! 이부분도 service 계층을 이용하도록 변경했습니다!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/domain/Comment.java:
```diff
@@ -1,18 +1,25 @@
 
 @Entity
+@Getter
+@NoArgsConstructor
```
기본생성자는 외부에서 사용할 수 없게끔 protected로 감춰주세요

kimnayeon0108 replied:
@ksundong 넵 엔티티 클래스 기본생성자 모두 protected로 변경했습니다~
이전에 JPA로 미션 했을때 엔티티 클래스의 기본생성자에 대해서 다른 학습동료들에게 여러번 설명 들었던거 같은데 생각이 잘 안나네요ㅠㅠ 이부분도 다시 학습해보겠습니다

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/domain/Comment.java:
```diff
@@ -1,18 +1,25 @@
     private String content;
     private String file;
     private LocalDateTime createdTime;
+    private boolean deleted;
```
isDeleted라는 네이밍을 더 추천드려요.

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/domain/CommentEmoji.java:
```diff
@@ -1,11 +1,17 @@
 @IdClass(CommentEmojiId.class)
 @Entity
+@Getter
+@AllArgsConstructor
+@NoArgsConstructor
```
저는 이런 생성자용 애노테이션은 손으로 써주는걸 좋아합니다.
실수할 여지도 없고, 치는데 얼마 걸리지도 않거든요!
특히 AllArgs는 주의해서 사용해야 하는 애노테이션입니다.

kimnayeon0108 replied:
@ksundong 넵~~! 롬복 어노테이션 사용보다 눈에 더 잘 띄는듯 하고 직접 생성자를 작성하는 편이 더 좋아보입니다. AllArgsConstructor 사용하는 부분 모두 직접 작성하도록 변경했습니다!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/domain/Emoji.java:
```diff
@@ -8,11 +10,13 @@
     private Long id;
 
     private String name;
+    private boolean deleted;
```
이모지가 삭제될 여지도 있는건가요?

ksundong replied:
그럼 뭔가 위의 설계가 이해가 가는 것 같아요.
서버가 이모지를 관리한다는건 이 팀만의 정책이군요 ㅎㅎ

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/domain/Issue.java:
```diff
@@ -1,23 +1,31 @@
 
     private LocalDateTime createdTime;
-    private boolean isOpen;
+    private boolean open;
+    private boolean deleted;
```
왜 네이밍을 바꿔주신걸까요?

kimnayeon0108 replied:
@ksundong 처음에 isOpen, isDeleted로 했다가 바꿔준 이유는 is가 붙는건 메소드명으로 사용하는게 더 적절하다고 생각해서 변경했습니다. 
그런데 디온이 기존처럼 isDeleted 방식을 추천하셔서 다시 변경했습니다~! 변수명이여도 is를 붙여서 사용하는게 더 컨벤션에 맞나 보네요!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/domain/Issue.java:
```diff
@@ -32,4 +40,60 @@
+            throw new IllegalStatusException("닫힌 이슈가 아닙니다.");
+        }
+        open = true;
+    }
```
👍

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/domain/Issue.java:
```diff
@@ -32,4 +40,60 @@
+        this.milestone = milestone;
+    }
+
+    public void delete() {
+        deleted = true;
+    }
```
Hibernate는 여러가지 soft delete 방식을 지원해요. 한 번 공부해보세요~

kimnayeon0108 replied:
@ksundong 넵 `@SQLDelete`와 `@Where` 어노테이션을 이용해서 soft delete를 구현했고 데이터 조회시 is_deleted 속성은 false인 것들만 조회되도록 했습니다!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/domain/Issue.java:
```diff
@@ -32,4 +40,60 @@
+        deleted = true;
+    }
+
+    public List<Comment> getComments() {
+        return comments.stream()
+                .filter(comment -> !comment.isDeleted())
+                .collect(Collectors.toList());
+    }
```
잘못하면 LazyLoding으로 인한 N + 1 쿼리가 발생할 수 있는 코드입니다.

kimnayeon0108 replied:
@ksundong 성능에 대한 부분을 여러군데 지적해주셔서 개선해보고자 지연 로딩과 영속성 전이에 대해서 학습하고 코드에 전반적으로 적용해 보았습니다. 연관관계를 맺은 필드들에 대해서 fetch 옵션으로 지연 로딩을 하도록 설정했고, 필요한 부분에 대해서 cascade 옵션을 ALL로 설정했습니다.

해당 메소드는 삭제 하였습니다~!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/domain/Issue.java:
```diff
@@ -32,4 +40,60 @@
+                .collect(Collectors.toList());
+    }
+
+    public void deleteMilestone() {
+        milestone = null;
+    }
```
milestone에서도 issue를 알텐데 그쪽에서는 바꿔주지 않아도 될까요?
연관관계 편의 메서드에 대해서 공부해보세요!

kimnayeon0108 replied:
@ksundong  김영한님 JPA 책을 보면서 학습 중인데, 외래키를 갖고 있는 쪽이 연관관계의 주인이 되어서 이 경우에는 Issue가 주인이 되고, 양방향 매핑에 대해서는 연관관계의 주인인 Issue 쪽에서만 변경해도 된다고 알고 있습니다...!

JPA 너무 어려워서 제가 착각할 수도 있을거 같아요ㅠㅠ 혹시 제가 잘못 알고 있다면 피드백 부탁드립니다...!🙏

kimnayeon0108 replied:
@ksundong 다시 찾아보니 양방향 매핑은 양쪽에서 관리해줘야 하는군요... 더 학습하고 수정하겠습니다...!

ksundong replied:
양방향 매핑은 원래 이론상 한 쪽에서만 관리해줘도 되긴하는데
문제는 객체지향 프로그래밍을 하다보면 둘이 엮이는 경우에 문제가 발생합니다.
그래서 연관관계 편의메서드라고 한 쪽에서 조작할 때 다른 한 쪽도 같이 변경되도록 많이 처리해줘요~

kimnayeon0108 replied:
@ksundong 아하! 영한님 책에서도 양쪽에서 처리해줘야 안전하다고 하더라구요~ 더 학습해보겠습니다~

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/domain/Label.java:
```diff
@@ -8,13 +13,37 @@
+        this.title = title;
+        this.content = content;
+        this.color = color;
+        this.deleted = false;
```
기본형은 기본값이 할당됩니다. 명시적으로 해주려고 했던 의도가 아닌이상 지정해주지 않아도 괜찮아요.

kimnayeon0108 replied:
@ksundong 넵 불필요한 코드 삭제했습니다!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/domain/Milestone.java:
```diff
@@ -9,16 +13,60 @@
+        this.open = true;
+    }
+
+    public int getTotalIssueCount() {
+        return (int) issues.stream()
+                .filter(issue -> !issue.isDeleted())
+                .count();
+    }
+
+    public int getOpenIssueCount() {
+        return (int) issues.stream()
+                .filter(issue -> !issue.isDeleted())
+                .filter(Issue::isOpen)
+                .count();
+    }
+
+    public int getClosedIssueCount() {
+        return (int) issues.stream()
+                .filter(issue -> !issue.isDeleted())
+                .filter(issue -> !issue.isOpen())
+                .count();
+    }
```
위험합니다~

ksundong replied:
N + 1 쿼리가 유도될 수 있어요!

kimnayeon0108 replied:
@ksundong 이 메소드들 전부 삭제하고 특정 milestone을 참조하는 Issue들의 개수를 조회하는 메소드를 IssueRepository에 추가해서 이를 이용하도록 로직 변경했습니다~

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/domain/Milestone.java:
```diff
@@ -9,16 +13,60 @@
+        deleted = true;
+        issues.stream()
+                .forEach(issue -> issue.deleteMilestone());
+    }
```
👍

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/dto/ApiResult.java:
```diff
@@ -0,0 +1,30 @@
+import lombok.Getter;
+
+@Getter
+public class ApiResult<T>{
```
```suggestion
public class ApiResult<T> {
```

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/dto/ApiResult.java:
```diff
@@ -0,0 +1,30 @@
+@Getter
+public class ApiResult<T>{
+    private T data;
+    private String error;
```
항상 error가 전달될 것을 가정하는 것은 별로인 것 같아요~

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/dto/ApiResult.java:
```diff
@@ -0,0 +1,30 @@
+        return new ApiResult<>(data, null);
+    }
+
+    public static ApiResult<?> fail(String errorMessage) {
+        return new ApiResult<>(null, errorMessage);
+    }
+
+    public static ApiResult<?> fail(Throwable cause) {
+        return fail(cause.getMessage());
+    }
```
제네릭 와일드카드의 사용은 별로 좋지 못한 것 같아요.

ehdrhelr replied:
fail시 보내주는 data가 없기 때문에 Void로 바꿔주었습니다.

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/dto/CommentEmojiRequest.java:
```diff
@@ -0,0 +1,11 @@
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
```
이거 정말 의미가 있을까요?
오히려 이는 Java를 잘 모른다는 인상을 줄 수 있어보입니다.

kimnayeon0108 replied:
@ksundong 기본 생성자 없이도 잘 동작하여 삭제해주었습니다! 

그런데 궁금한 점이, 이전 미션 진행시 Request DTO 클래스에 기본생성자가 없으면 에러가 발생해서 찾아본적이 있는데 역직렬화 하는 과정에서 어떤 생성자를 갖고 있는지 모르기 때문에 기본생성자를 호출하기 때문에 에러가 발생한다는 것을 학습한 적이 있습니다. 
그런데 왜 JPA 사용하는 현재 시점에서는 기본생성자가 없어도 되는 건지... 궁금합니다...!

kimnayeon0108 replied:
@ksundong 앗... 맞습니다 불필요한 부분 전부 삭제했습니다...!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/dto/LabelCountResponse.java:
```diff
@@ -0,0 +1,11 @@
+@Getter
+public class LabelCountResponse {
+
+    private int labels;
```
labelCount가 더 맞겠죠?

kimnayeon0108 replied:
@ksundong 넵 변경했습니다~!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/dto/issue/AbstractIssueResponse.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+@AllArgsConstructor
```
추상클래스의 public 생성자는 코드스멜이에요~

ksundong replied:
그리고 Abstract같은 것은 저는 별로 선호하는 네이밍이 아니지만 일단 넘어가봅시다.

kimnayeon0108 replied:
@ksundong protected 로 접근제어자 변경했습니다! 추상 클래스의 네이밍에 대해서도 더 학습해볼게요!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/dto/issue/DetailIssueResponse.java:
```diff
@@ -0,0 +1,33 @@
+@Getter
+public class DetailIssueResponse extends AbstractIssueResponse {
+
+    private Optional<DetailIssueMilestoneResponse> milestone;
```
리턴 값에 Optional을 사용해주는건 별로 좋지 않은 것 같아요.

kimnayeon0108 replied:
@ksundong 이슈에 해당하는 마일스톤이 없으면 null 값을 넘겨주기로 해서 null을 직접 다루기보다는 Optional을 사용하는게 더 나을거 같다고 생각했습니다....! 
말씀하신대로 Optional 제거했습니다~ 이런 경우에는 null을 반환하기 보다는 빈 String 값을 넘겨주면 될지... 설계를 어떤식으로 하면 좋을지 조언주시면 감사하겠습니다..🙇‍♀️🙇‍♀️

ksundong replied:
Optional을 사용하는 것은 우리 내부 코드에서만 한정짓고, Presentation Layer에서는 알지 않는 편이 좋은 것 같아요.
저는 한 번도 Optional 객체를 반환해본적이 없어서 어떻게 값이 전달될지 잘 모르겠네요.

저는 되도록이면 null값을 포함해서 리턴할 것인지, 아니면 제외할 것인지를 개발팀에서 정해야 한다고 생각합니다!

kimnayeon0108 replied:
@ksundong 아 그렇군요...! 감사합니다~ null값을 다루기보다는 아예 제외하는게 현재로서는 더 좋은 방법처럼 보이네요 다음부터 적용해봐야겠습니다!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/exception/CommentNotFoundException.java:
```diff
@@ -0,0 +1,8 @@
+package team02.issue_tracker.exception;
+
+public class CommentNotFoundException extends NotFoundException{
```
```suggestion
public class CommentNotFoundException extends NotFoundException {
```

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/exception_handler/GlobalExceptionHanlder.java:
```diff
@@ -0,0 +1,39 @@
+
+    @ExceptionHandler(IncorrectTokenTypeException.class)
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public ApiResult incorrectTokenTypeException(IncorrectTokenTypeException e) {
```
어떤 Type인가요? Bearer 같은걸까요?

ehdrhelr replied:
넵 현재는 토큰타입이 "Bearer"로 시작하지 않으면 발생하는 예외입니다.

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/exception_handler/GlobalExceptionHanlder.java:
```diff
@@ -0,0 +1,39 @@
+
+    @ExceptionHandler(IllegalStatusException.class)
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public ApiResult illegalIssueStatusException(IllegalStatusException e) {
```
리턴타입이 다 Raw 타입이네요.
제네릭 Raw 타입의 사용은 항상 지양해주세요.

ehdrhelr replied:
넵 에러발생 시 반환되는 data는 없어서 Void로 바꾸면 될 것 같습니다~!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/filter/CorsFilter.java:
```diff
@@ -0,0 +1,29 @@
+public class CorsFilter implements Filter {
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+        HttpServletResponse httpServletResponse = (HttpServletResponse) response;
+        httpServletResponse.setHeader("Access-Control-Allow-Origin", "*");
+        httpServletResponse.setHeader("Access-Control-Allow-Credentials", "true");
+        chain.doFilter(request, response);
+    }
```
왜 Filter를 사용해주셨나요?
그리고 Origin을 *로 지정하면 어떻게 동작할까요?

kimnayeon0108 replied:
@ksundong Filter를 사용한 이유는 addCorsMapping으로 설정을 했는데도 프론트엔드에서 CORS 문제해결이 안된다고 해서 임시방편으로 헤더를 명시적으로 달아주기 위해서 사용했습니다. 
*로 지정하면 모든 Origin에서 접속할 수 있어서 저희의 의도와는 맞지 않은거 같고 프론트 서버의 Origin인 `http://localhost:3000` 으로 설정해주는것이 좋을거 같습니다. 
CORS 동작 원리가 여러가지 있던데 이부분에 대해 더 학습하고 해결해야할 듯 합니다. 그렇다면 Filter는 필요 없을 듯 합니다...!

ksundong replied:
WebConfigurer에서도 설정가능합니다. 한 번 시도해보세요.
*로 지정하는 동작방식은 좀 더 알아보세요. 제대로 동작하지 않을 가능성이 높습니다.
화이팅~

kimnayeon0108 replied:
@ksundong 저랑 시온이랑 오늘 CORS에 대해 자세히 학습해보고, preflight 요청에 대해서는 Option 메소드로 요청이 들어온다는 사실을 알게 되었습니다~ 

```
@Override
    public void addCorsMappings(CorsRegistry registry) {
        registry.addMapping("/api/**")
                .allowedMethods("*")
                .allowedOrigins("http://localhost:3000");
    }
```
라고 allowedMethods에 전체 HTTP 메소드를 허용하도록 수정하고 필터는 제거했습니다!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/oauth/dto/GithubUserProfile.java:
```diff
@@ -0,0 +1,27 @@
+    @JsonSetter("avatar_url")
+    private String avatarUrl;
+
+    public User becomeUser() {
```
Override 애노테이션이 누락되었네요.
조심해주세요.

ehdrhelr replied:
추가해주었습니다 🥲

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/oauth/dto/SocialProfile.java:
```diff
@@ -0,0 +1,7 @@
+import team02.issue_tracker.domain.User;
+
+public interface SocialProfile {
+    User becomeUser();
```
become 말고 다른 동사를 사용해보는건 어떨까요?

ehdrhelr replied:
저는 become이 썩 마음에 드는데 컨벤션(?)에 맞는 더 좋은 메서드명이 있을까요? ㅠㅠ

ksundong replied:
사실 become은 스스로 되는 느낌이긴해서 나쁘지 않은것 같네요!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/oauth/interceptor/JwtInterceptor.java:
```diff
@@ -0,0 +1,67 @@
+@Component
+public class JwtInterceptor implements HandlerInterceptor {
+
+    private final int JWT = 1;
```
무슨의미일까요?

ehdrhelr replied:
header에 담긴 "Bearer JWT"를 띄어쓰기로 split해서 String 배열로 담는데 index 0은 토큰타입이고 1은 JWT입니다.

ksundong replied:
그러면 JWT_INDEX 같은 방식은 어떨까요?

ehdrhelr replied:
INDEX를 추가해주는게 더 의미상 맞을 것 같습니다!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/oauth/interceptor/JwtInterceptor.java:
```diff
@@ -0,0 +1,67 @@
+
+    private final int JWT = 1;
+
+    private final JwtUtils jwtUtils;
+
+    public JwtInterceptor(JwtUtils jwtUtils) {
+        this.jwtUtils = jwtUtils;
+    }
```
Utils는 저같은 경우에는 static 한 공통로직으로 쓰일 수 있는 메서드들을 관리하는 클래스로 사용하고
이를 의존성 관리하지는 않습니다.

ehdrhelr replied:
말씀 들어보니 그렇게 하는 편이 더 나은 것 같습니다!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/oauth/interceptor/JwtInterceptor.java:
```diff
@@ -0,0 +1,67 @@
+                && ((HandlerMethod) handler).hasMethodAnnotation(LoginRequired.class);
+    }
+
+    private void setUserIdOn(HttpServletRequest request) {
+        String jwt = extractJwtFromHeader(request);
+        Long userId = decodeJwtToUserId(jwt);
+        request.setAttribute("user_id", userId);
+    }
+
+    private String extractJwtFromHeader(HttpServletRequest request) {
+        String authorizationInHeader = request.getHeader("Authorization");
+        verify(authorizationInHeader);
+        String[] tokenTypeAndJwt = authorizationInHeader.split(" ");
+        return tokenTypeAndJwt[JWT];
+    }
+
+    private void verify(String authorizationInHeader) {
+        if (authorizationInHeader == null || authorizationInHeader.isEmpty()) {
+            throw new JwtNotFoundException();
+        }
+        if (!authorizationInHeader.startsWith("Bearer")) {
+            throw new IncorrectTokenTypeException();
+        }
+    }
+
+    private Long decodeJwtToUserId(String jwt) {
+        DecodedJWT decodedJwt = jwtUtils.decode(jwt);
+        return decodedJwt.getClaim("user_id").asLong();
+    }
```
private이 너무 많은것도 좋은 패턴은 아닙니다.
최대한 줄여보세요. 클래스로 나누거나 하는 방법이 있겠죠?

ehdrhelr replied:
JwtFinder 클래스를 만들어서 header에서 jwt를 추출하게하고, JwtFinder를 만들어서 decode하게 수정했습니다.
더불어 JwtUtils를 JwtFactory로 이름을 바꾸고 Jwt를 생성하는 역할만 수행하도록 변경하였습니다!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/service/IssueService.java:
```diff
@@ -0,0 +1,146 @@
+        return new DetailIssueResponse(issue);
+    }
+
+    public void addIssue(IssueRequest issueRequest, Long userId) {
+        User writer = userService.findOne(userId);
+
+        Issue issue = makeIssue(issueRequest, writer);
+        commentService.makeComment(issueRequest, writer, issue);
+        labelService.makeIssueLabels(issue, issueRequest.getLabelIds());
+        userService.makeIssueAssignees(issue, issueRequest.getAssigneeIds());
+    }
```
이런경우에 필요한 애노테이션이 있을겁니다.
한 번 알아보세요!

kimnayeon0108 replied:
@ksundong  기존에 makeComment, makeIssueLabels, makeIssueAssignees 메소드들의 기능을 변경했습니다.
기존에는 생성하고 save까지 하도록 했었는데, 현재는 save 하지 않고 Comment, IssueLabel, IssueAssignees들만 생성하도록 변경했습니다. 
그리고 Issue 클래스에서 관련된 필드들에 cascade 옵션을 주고, 새로 생성된 Issue 객체에 코멘트와 그 외에 것들을 추가해주는 메소드들(addComment, addIssueLabels 등) 을 생성하고 사용해서 Issue만 save하면 다른 객체들도 저장되도록 변경했습니다! 

이런경우에 필요한 어노테이션이 Transaction을 말씀하시는거 같아서 이것도 적용시켰습니다...!

ksundong commented on BE/issue_tracker/src/main/java/team02/issue_tracker/service/UserService.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    public List<IssueAssignee> makeIssueAssignees(Issue issue, List<Long> assigneeIds) {
+        List<IssueAssignee> issueAssignees = new ArrayList<>();
+
+        assigneeIds.forEach(assigneeId -> {
+                    User assignee = userRepository.findById(assigneeId).orElseThrow(UserNotFoundException::new);
+                    issueAssignees.add(new IssueAssignee(issue, assignee));
+                });
```
stream을 잘 활용해보면 되지 않을까요?

kimnayeon0108 replied:
@ksundong 아직 스트림 사용이 미숙해서....ㅠㅠ 이부분도 아래와 같이 변경했는데 올바르게 한건지... 피드백 부탁드립니다..!!!

```
 public List<IssueAssignee> makeIssueAssignees(Issue issue, List<Long> assigneeIds) {
        return assigneeIds.stream()
                .map(assigneeId -> {
                    User assignee = userRepository.findById(assigneeId).orElseThrow(UserNotFoundException::new);
                    return new IssueAssignee(issue, assignee);
                }).collect(Collectors.toList());
    }
```

ksundong replied:
자 그러면 매번 DB 호출하는 부분이 눈에 띌텐데요.
어떻게 처리 할지는 고민해보시고, 고치느라 수고하셨습니다!

kimnayeon0108 replied:
@ksundong 헉 그렇군요.... 더 고민해보겠습니다 감사합니다~

ksundong reviewed (approved):
안녕하세요! 연, 시온!! 리뷰어 Dion입니다. 벌써 마지막이 다가오고 있네요. 앞으로 얼마 안남은만큼 더 최선을 다해주시길 바랍니다.
개인적으로 컨디션이 조금 많이 안좋아서 리뷰가 늦어져서 정말 죄송합니다. 운동을 너무 힘내서 해도 문제가 되네요. ㅋㅋ

리뷰를 하면서 개인적으로 아쉬웠던 점은, 규모가 커지다보니 필연적으로 일정부분 관성적인 코드가 일부 보였습니다.
어쩔 수 없다는 것을 알면서도 아쉽네요.

그리고, Repository Test를 수행하시기 보다는 본질적인 Domain 테스트에 시간을 들여보셨으면 더 좋았을 것 같습니다.
제 개인적으로는 Repository Test는 직접 작성한 구현코드가 없는 한 할 필요가 없다고 생각됩니다.
마치 스프링 개발자와 Hibernate 개발자들이 이미 테스트 한 것을 또 테스트하는 느낌이라고 생각해보시면 될 것 같아요.

우리가 필요한 것은 도메인 로직이 정상적으로 수행될 것인가?라는 질문에 대한 답을 주는 것입니다.
가령 게시글 작성자가 내용을 수정했다면 그 내용이 수정되었는지를 DB에 질의하지 않고도 우리는 알 수 있어야 합니다.
그것이 우리가 JPA라는 ORM 기술을 사용하는 이유라고도 생각됩니다.

이를 가장 잘 이해하기 위해서는 우리가 작성한 서비스가 빈약하고, 도메인이 풍성한지를 보면 됩니다.
한 번 살펴보시면, 서비스가 지나치게 로직을 많이 담고 있다는 사실도 아실 수 있을거에요.

질문하신 내용은 create-drop 혹은 update로 설정하면 되지 않을까요?
사실 실무에서는 항상 validate로 동작시키거나 none으로 두기 때문에 해당 이슈에 대한 해결방안을 알고있진 않네요.
제가 알기론 create는 생성만 시도하고, drop은 하지 않는 것으로 알기 때문에 한 번 create-drop을 시도해보시길 바랍니다.

고생 많으셨습니다!

kimnayeon0108 committed:
refactor : CommonController url 변경

kimnayeon0108 committed:
chore : gitignore 업데이트

kimnayeon0108 committed:
refactor : 개수 조회하는 메소드 변경

kimnayeon0108 committed:
refactor : 유저 조회하는 기능 코드 리팩토링

kimnayeon0108 committed:
refactor : 엔티티 기본생성자 접근제어자 protected로 변경

kimnayeon0108 committed:
refactor : 모든 인자를 갖는 생성자 롬복 기능 대신 명시적으로 수정

kimnayeon0108 committed:
chore : 불필요한 코드 삭제

kimnayeon0108 committed:
refactor : boolean 필드명 변경에 따른 메소드명과 코드 변경

Commit:
[#46] refactor : Javadoc사용, 포인트컷 메서드, 어노테이션화

Commit:
[#46] refactor : ApiResult 제네릭 지정

Commit:
[#46] style : 누락된 @Override 추가

Commit:
[#46] refactor : JwtFactory, JwtFinder, JwtDecoder로 업무 분담

kimnayeon0108 committed:
Merge pull request #47 from Sonjh1306/BE/refactor/feedback-s

[BE] 코드리뷰 받은대로 수정하였습니다.

kimnayeon0108 committed:
refactor : hibernate의 soft delete 방식 적용

SQLDelete와 Where 어노테이션 추가
Repository 메소드명 변경

kimnayeon0108 committed:
feat : IssueLabel 지연로딩 옵션 추가

kimnayeon0108 committed:
refactor : Repository 메소드 수정에 따른 코드 수정

kimnayeon0108 committed:
refactor : Repository 메소드 수정에 따른 코드 수정

kimnayeon0108 committed:
refactor : 이슈 추가하기 기능 리팩토링

kimnayeon0108 committed:
refactor : IssueLabel 수정하기, IssueAssignee 수정하기 기능 리팩토링

kimnayeon0108 committed:
feat : 이슈 삭제 기능 수정

Issue와 Comment에서 one to many 관계의 필드에 영속성 전이 옵션 추가

kimnayeon0108 committed:
feat : 레이블 삭제 기능 수정

Label에서 one to many 관계의 필드에 영속성 전이 옵션 추가

kimnayeon0108 committed:
chore: 코드 줄맞춤, 불필요한 메소드 삭제

kimnayeon0108 committed:
refactor: 메소드 접근제어자 변경

kimnayeon0108 committed:
feat : 마일스톤 전체 조회 기능 수정

milestone에서 getIssues()를 하지 않고, IssueRepository의 메소드를 통해서 이슈의 개수를 조회하도록 변경
N+1 문제 방지를 위한 변경

kimnayeon0108 committed:
feat : 이슈 개별 조회 기능 수정

이슈 개별 조회시 마일스톤에 해당하는 이슈의 개수 반환하는 로직 수정
milestone을 통해 issue를 접근하는 기존 로직에서 IssueRepository를 통해 개수를 조회하도록 변경

kimnayeon0108 committed:
refactor : 이슈 전체 조회 DTO 반환시 옵셔널 제거

kimnayeon0108 committed:
refactor : 마일스톤 삭제 기능 수정

kimnayeon0108 committed:
test : Repository의 메소드 기능을 확인하는 테스트 코드 삭제

kimnayeon0108 committed:
refactor: 마일스톤 삭제시 이슈의 마일스톤 값 null 처리 하는 로직 수정

kimnayeon0108 committed:
refactor: 불필요한 기본생성자 삭제, 모든 매개변수를 갖는 생성자 명시적으로 작성

kimnayeon0108 committed:
refactor: LabelCountResponse 필드명 변경

kimnayeon0108 committed:
refactor: AbstractIssueResponse 생성자 접근제어자 변경

kimnayeon0108 committed:
refactor: 불필요한 기본생성자 삭제

kimnayeon0108 committed:
refactor: AbstractIssueResponse 클래스의 불필요한 생성자 삭제, 생성자의 접근제어자 protected로 변경

kimnayeon0108 committed:
chore: 코드 줄맞춤

kimnayeon0108 committed:
refactor: AllArgsConstructor 어노테이션 사용 대신 직접 생성자 작성하도록 변경

kimnayeon0108 committed:
chore: build.gradle 수정

kimnayeon0108 committed:
feat: User 엔티티의 id 생성 전략 설정

kimnayeon0108 committed:
refactor: 이슈 추가 기능 리팩토링

kimnayeon0108 committed:
chore: 코드 줄맞춤

kimnayeon0108 commented:
@ksundong 말씀해주신 사항 학습하고 적용하느라 피드백 반영이 좀 늦어진거 같습니다...🥲 
몸은 좀 괜찮아지셨나요ㅠㅠ 아프지마세요...!💪💪  시간 나실때 피드백 반영사항 확인 부탁드립니다!

Commit:
Merge pull request #48 from Sonjh1306/BE/refactor/feedback-y

[#46] 코드리뷰 피드백 반영
